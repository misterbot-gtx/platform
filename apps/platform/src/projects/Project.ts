import Model, { ModelParams } from '../core/Model'

export default class Project extends Model {
    organization_id!: number
    name!: string
    description?: string
    deleted_at?: Date
    locale!: string
    timezone!: string
    text_opt_out_message?: string
    text_help_message?: string
    link_wrap_email?: boolean
    link_wrap_push?: boolean
}

export type ProjectParams = Omit<Project, ModelParams | 'deleted_at' | 'organization_id'>

export const projectRoles = [
    'support',
    'editor',
    'admin',
] as const

export type ProjectRole = (typeof projectRoles)[number]
