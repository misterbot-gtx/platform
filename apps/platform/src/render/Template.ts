import Render, { Variables, Wrap } from '.'
import { Webhook } from '../providers/webhook/Webhook'
import { ChannelType } from '../config/channels'
import Model, { ModelParams } from '../core/Model'
import { isValid, IsValidSchema } from '../core/validate'
import { Email, NamedEmail } from '../providers/email/Email'
import { htmlToText } from 'html-to-text'
import { paramsToEncodedLink } from './LinkService'

export default class Template extends Model {
    project_id!: number
    campaign_id!: number
    type!: ChannelType
    data!: Record<string, any>
    locale!: string

    static tableName = 'templates'

    static jsonAttributes = ['data']

    map(): TemplateType {
        const json = this as any
        if (this.type === 'email') {
            return EmailTemplate.fromJson(json)
        } else if (this.type === 'text') {
            return TextTemplate.fromJson(json)
        } else if (this.type === 'push') {
            return PushTemplate.fromJson(json)
        }
        return WebhookTemplate.fromJson(json)
    }

    validate(): IsValidSchema {
        return [true, undefined]
    }

    requiredErrors(...fields: string[]) {
        const errors: Record<string, string> = {}
        for (const field of fields) {
            errors[field] = `The \`${field}\` field on the \`${this.locale}\` template is missing and is required.`
        }
        return errors
    }
}

export type TemplateParams = Omit<Template, ModelParams | 'map' | 'screenshotUrl' | 'validate' | 'requiredErrors'>
export type TemplateUpdateParams = Pick<Template, 'type' | 'data'>
export type TemplateType = EmailTemplate | TextTemplate | PushTemplate | WebhookTemplate

type CompiledEmail = Omit<Email, 'to' | 'headers'> & { preheader?: string }

export class EmailTemplate extends Template {
    declare type: 'email'
    from!: string | NamedEmail
    cc?: string
    bcc?: string
    reply_to?: string
    subject!: string
    preheader?: string
    editor!: 'code' | 'visual'
    text?: string
    html!: string
    mjml?: string

    parseJson(json: any) {
        super.parseJson(json)

        this.from = json?.data.from
        this.cc = json?.data.cc
        this.bcc = json?.data.bcc
        this.reply_to = json?.data.reply_to
        this.subject = json?.data.subject ?? ''
        this.preheader = json?.data.preheader
        this.editor = json?.data.type ?? 'code'
        this.text = json?.data.text
        this.html = json?.data.html ?? ''
        this.mjml = json?.data.mjml
    }

    compile(variables: Variables): CompiledEmail {
        const html = Render(this.html, variables)
        const preheader = this.preheader ? Render(this.preheader, variables) : undefined
        const email: CompiledEmail = {
            subject: Render(this.subject, variables),
            from: typeof this.from === 'string'
                ? Render(this.from, variables)
                : {
                    name: Render(this.from?.name ?? '', variables),
                    address: Render(this.from?.address ?? '', variables),
                },
            html: Wrap({ html, preheader, variables }), // Add link and open tracking

            // If the text copy has been left empty, generate from HTML
            text: this.text !== undefined && this.text !== ''
                ? Render(this.text, variables)
                : htmlToText(html),
        }

        if (preheader) email.preheader = preheader
        if (this.reply_to) email.reply_to = Render(this.reply_to, variables)
        if (this.cc) email.cc = Render(this.cc, variables)
        if (this.bcc) email.bcc = Render(this.bcc, variables)
        return email
    }

    validate() {
        return isValid({
            type: 'object',
            required: ['from', 'subject', 'html'],
            properties: {
                from: {
                    type: 'object',
                    required: ['address'],
                    properties: {
                        name: { type: 'string', nullable: true },
                        address: { type: 'string' },
                    },
                },
                subject: { type: 'string' },
                text: { type: 'string' },
                html: { type: 'string' },
            },
            additionalProperties: true,
            errorMessage: {
                required: this.requiredErrors('from', 'subject', 'text', 'html'),
            },
        }, this.data)
    }
}

export interface CompiledText {
    text: string
}

export class TextTemplate extends Template {
    declare type: 'text'
    text!: string

    parseJson(json: any) {
        super.parseJson(json)

        this.text = json?.data.text
    }

    compile(variables: Variables): CompiledText {
        return { text: Render(this.text ?? '', variables) }
    }

    validate() {
        return isValid({
            type: 'object',
            required: ['text'],
            properties: {
                text: { type: 'string' },
            },
            errorMessage: {
                required: this.requiredErrors('text'),
            },
        }, this.data)
    }
}

export interface CompiledPush {
    title: string
    topic: string
    body: string
    custom: Record<string, any>
}

export class PushTemplate extends Template {
    declare type: 'push'
    title!: string
    topic!: string
    body!: string
    url!: string
    custom!: Record<string, any>

    parseJson(json: any) {
        super.parseJson(json)

        this.title = json?.data.title
        this.topic = json?.data.topic
        this.body = json?.data.body
        this.url = json?.data.url
        this.custom = json?.data.custom ?? {}
    }

    compile(variables: Variables): CompiledPush {
        const custom = Object.keys(this.custom).reduce((body, key) => {
            body[key] = Render(this.custom[key], variables)
            return body
        }, {} as Record<string, any>)

        const url = this.compileUrl(variables)

        return {
            topic: this.topic,
            title: Render(this.title, variables),
            body: Render(this.body, variables),
            custom: { ...custom, url },
        }
    }

    compileUrl(variables: Variables) {
        if (this.url === undefined) {
            return undefined
        }

        const { project, user, context } = variables
        const renderedUrl = Render(this.url, variables)

        return project.link_wrap_push
            ? paramsToEncodedLink({
                userId: user.id,
                campaignId: context.campaign_id,
                referenceId: context.reference_id,
                redirect: renderedUrl,
                path: 'c',
            })
            : renderedUrl
    }

    validate() {
        return isValid({
            type: 'object',
            required: ['title', 'topic', 'body'],
            properties: {
                title: { type: 'string' },
                topic: { type: 'string' },
                body: { type: 'string' },
                url: { type: 'string', nullable: true },
            },
            additionalProperties: true,
            errorMessage: {
                required: this.requiredErrors('title', 'topic', 'body'),
            },
        }, this.data)
    }
}

export class WebhookTemplate extends Template {
    declare type: 'webhook'
    method!: 'DELETE' | 'GET' | 'PATCH' | 'POST' | 'PUT'
    endpoint!: string
    body!: Record<string, any>
    headers: Record<string, string> = {}

    parseJson(json: any) {
        super.parseJson(json)

        this.method = json?.data.method
        this.endpoint = json?.data.endpoint
        this.body = json?.data.body
        this.headers = json?.data.headers || {}
    }

    compile(variables: Variables): Webhook {
        const headers = Object.keys(this.headers ?? {}).reduce((headers, key) => {
            headers[key] = Render(this.headers[key], variables)
            return headers
        }, {} as Record<string, string>)

        const body = Object.keys(this.body ?? {}).reduce((body, key) => {
            body[key] = Render(this.body[key], variables)
            return body
        }, {} as Record<string, any>)

        const endpoint = Render(this.endpoint, variables)
        const method = this.method
        return {
            endpoint,
            method,
            headers,
            body,
        }
    }

    validate() {
        return isValid({
            type: 'object',
            required: ['method', 'endpoint'],
            properties: {
                method: { type: 'string' },
                endpoint: { type: 'string', nullable: true },
            },
            additionalProperties: true,
            errorMessage: {
                required: this.requiredErrors('method', 'endpoint'),
            },
        }, this.data)
    }
}
