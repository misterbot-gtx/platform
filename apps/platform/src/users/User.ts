import { ClientIdentity } from '../client/Client'
import Model, { ModelParams } from '../core/Model'
import parsePhoneNumber from 'libphonenumber-js'

export interface TemplateUser extends Record<string, any> {
    id: string
    email?: string
    phone?: string
}

export interface UserAttribute {
    id: number
    user_id: number
    key: string
    value: any
}

export interface Device {
    device_id: string
    token?: string
    os?: string
    os_version?: string
    model?: string
    app_build?: string
    app_version?: string
}

export type DeviceParams = Omit<Device, ModelParams> & ClientIdentity

interface PushEnabledDevice extends Device {
    token: string
}

export class User extends Model {
    project_id!: number
    anonymous_id!: string
    external_id!: string
    email?: string
    phone?: string
    devices?: Device[]
    data!: Record<string, any> // first_name, last_name live in data
    timezone?: string
    locale?: string

    static jsonAttributes = ['data', 'devices']
    static virtualAttributes = ['firstName', 'lastName', 'fullName']

    flatten(): TemplateUser {
        return {
            ...this.data,
            email: this.email,
            phone: this.phone,
            id: this.external_id,
            external_id: this.external_id,
            created_at: this.created_at,
            locale: this.locale,
            timezone: this.timezone,
        }
    }

    get pushEnabledDevices(): PushEnabledDevice[] {
        return this.devices?.filter(device => device.token != null) as PushEnabledDevice[] ?? []
    }

    get fullName() {
        // Handle case were user has a full name attribute in data
        const fullName = this.data.full_name ?? this.data.fullName
        if (fullName) return fullName

        // If no attribute exists, combine first and last name
        const parts: string[] = []
        if (this.firstName) {
            parts.push(this.firstName)
        }
        if (this.lastName) {
            parts.push(this.lastName)
        }
        return parts.join(' ') || null
    }

    get firstName(): string {
        return this.data.first_name ?? this.data.firstName ?? this.data.name
    }

    get lastName(): string {
        return this.data.last_name ?? this.data.lastName ?? this.data.surname
    }

    toJSON() {
        const json = super.toJSON()

        if (this.phone) {
            const parsedNumber = parsePhoneNumber(this.phone)
            if (parsedNumber) {
                json.phone = parsedNumber.formatInternational()
            }
        }

        return json
    }
}

export type UserParams = Partial<Pick<User, 'email' | 'phone' | 'timezone' |'locale' | 'data'>> & ClientIdentity
export type UserInternalParams = Partial<Pick<User, 'email' | 'phone' | 'timezone' |'locale' | 'created_at' | 'data'>> & ClientIdentity
