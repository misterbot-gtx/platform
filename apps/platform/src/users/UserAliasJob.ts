import { Job } from '../queue'
import { aliasUser } from './UserRepository'
import { ClientAliasParams } from '../client/Client'

type UserAliasTrigger = ClientAliasParams & {
    project_id: number
}

export default class UserAliasJob extends Job {
    static $name = 'user_alias'

    static from(data: UserAliasTrigger): UserAliasJob {
        return new this(data)
    }

    static async handler({ project_id, ...alias }: UserAliasTrigger) {
        await aliasUser(project_id, alias)
    }
}
