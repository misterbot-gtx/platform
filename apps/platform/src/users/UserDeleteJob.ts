import App from '../app'
import { Job } from '../queue'

interface UserDeleteTrigger {
    project_id: number
    external_id: string
}

export default class UserDeleteJob extends Job {
    static $name = 'user_delete'

    static from(data: UserDeleteTrigger): UserDeleteJob {
        return new this(data)
    }

    static async handler({ project_id, external_id }: UserDeleteTrigger) {

        await App.main.db.transaction(async trx =>
            trx('users')
                .where({ project_id, external_id })
                .delete(),
        )
    }
}
