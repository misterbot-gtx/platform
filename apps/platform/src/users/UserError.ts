export default {
    NotFound: {
        message: 'No user matching the given criteria was found.',
        code: 3000,
    },
}
