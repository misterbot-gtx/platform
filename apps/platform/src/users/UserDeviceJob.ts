import { Job } from '../queue'
import { saveDevice } from './UserRepository'
import { DeviceParams } from './User'

type UserDeviceTrigger = DeviceParams & {
    project_id: number
}

export default class UserDeviceJob extends Job {
    static $name = 'user_register_device'

    static from(data: UserDeviceTrigger): UserDeviceJob {
        return new this(data)
    }

    static async handler({ project_id, ...device }: UserDeviceTrigger, job: UserDeviceJob) {
        const attempts = job.options.attempts ?? 1
        const attemptsMade = job.state.attemptsMade ?? 0

        try {
            await saveDevice(project_id, device)
        } catch (error) {
            if (attemptsMade < (attempts - 1)) throw error
        }
    }
}
