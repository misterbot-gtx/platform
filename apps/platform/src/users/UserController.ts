import Router from '@koa/router'
import App from '../app'
import { ProjectState } from '../auth/AuthMiddleware'
import UserDeleteJob from './UserDeleteJob'
import UserPatchJob from './UserPatchJob'
import { JSONSchemaType, validate } from '../core/validate'
import { User, UserParams } from './User'
import { extractQueryParams } from '../utilities'
import { searchParamsSchema, SearchSchema } from '../core/searchParams'
import { getUser, getUserFromContext, pagedUsers } from './UserRepository'
import { getUserLists } from '../lists/ListService'
import { getUserSubscriptions, toggleSubscription } from '../subscriptions/SubscriptionService'
import { SubscriptionState } from '../subscriptions/Subscription'
import { getUserEvents } from './UserEventRepository'
import { projectRoleMiddleware } from '../projects/ProjectService'
import { pagedEntrancesByUser } from '../journey/JourneyRepository'

const router = new Router<
    ProjectState & { user?: User }
>({
    prefix: '/users',
})

router.get('/', async ctx => {
    const searchSchema = SearchSchema('usersSearchSchema', {
        sort: 'id',
        direction: 'desc',
    })
    const params = extractQueryParams(ctx.query, searchSchema)
    ctx.body = await pagedUsers(params, ctx.state.project.id)
})

const patchUsersRequest: JSONSchemaType<UserParams[]> = {
    $id: 'patchUsers',
    type: 'array',
    items: {
        anyOf: [{
            type: 'object',
            required: ['anonymous_id'],
            properties: {
                anonymous_id: {
                    type: 'string',
                },
                external_id: {
                    type: 'string',
                    nullable: true,
                },
                email: {
                    type: 'string',
                    nullable: true,
                },
                phone: {
                    type: 'string',
                    nullable: true,
                },
                timezone: {
                    type: 'string',
                    nullable: true,
                },
                locale: {
                    type: 'string',
                    nullable: true,
                },
                data: {
                    type: 'object',
                    nullable: true,
                    additionalProperties: true,
                },
            },
        },
        {
            type: 'object',
            required: ['external_id'],
            properties: {
                anonymous_id: {
                    type: 'string',
                    nullable: true,
                },
                external_id: {
                    type: 'string',
                },
                email: {
                    type: 'string',
                    nullable: true,
                },
                phone: {
                    type: 'string',
                    nullable: true,
                },
                timezone: {
                    type: 'string',
                    nullable: true,
                },
                locale: {
                    type: 'string',
                    nullable: true,
                },
                data: {
                    type: 'object',
                    nullable: true,
                    additionalProperties: true,
                },
            },
        }],
    },
    minItems: 1,
}
router.patch('/', projectRoleMiddleware('editor'), async ctx => {
    const users = validate(patchUsersRequest, ctx.request.body)

    for (const user of users) {
        await App.main.queue.enqueue(UserPatchJob.from({
            project_id: ctx.state.project.id,
            user,
        }))
    }

    ctx.status = 204
    ctx.body = ''
})

const deleteUsersRequest: JSONSchemaType<string[]> = {
    $id: 'deleteUsers',
    type: 'array',
    items: {
        type: 'string',
    },
    minItems: 1,
}
router.delete('/', projectRoleMiddleware('editor'), async ctx => {

    let userIds = ctx.request.query.user_id || []
    if (!Array.isArray(userIds)) userIds = userIds.length ? [userIds] : []

    userIds = validate(deleteUsersRequest, userIds)

    for (const externalId of userIds) {
        await App.main.queue.enqueue(UserDeleteJob.from({
            project_id: ctx.state.project.id,
            external_id: externalId,
        }))
    }

    ctx.status = 204
    ctx.body = ''
})

router.param('userId', async (value, ctx, next) => {
    ctx.state.user = await getUserFromContext(ctx)
    if (!ctx.state.user) {
        ctx.throw(404)
        return
    }
    return await next()
})

router.get('/:userId', async ctx => {
    ctx.body = ctx.state.user
})

router.delete('/:userId', projectRoleMiddleware('editor'), async ctx => {
    await App.main.queue.enqueue(UserDeleteJob.from({
        project_id: ctx.state.project.id,
        external_id: ctx.state.user!.external_id,
    }))

    ctx.status = 204
    ctx.body = ''
})

router.get('/:userId/lists', async ctx => {
    const params = extractQueryParams(ctx.query, searchParamsSchema)
    ctx.body = await getUserLists(ctx.state.user!.id, params, ctx.state.project.id)
})

router.get('/:userId/events', async ctx => {
    const searchSchema = SearchSchema('userEventSearchSchema', {
        sort: 'id',
        direction: 'desc',
    })
    const params = extractQueryParams(ctx.query, searchSchema)
    ctx.body = await getUserEvents(ctx.state.user!.id, params, ctx.state.project.id)
})

router.get('/:userId/subscriptions', async ctx => {
    const params = extractQueryParams(ctx.query, searchParamsSchema)
    ctx.body = await getUserSubscriptions(ctx.state.user!.id, params, ctx.state.project.id)
})

router.patch('/:userId/subscriptions', async ctx => {
    const subscriptions = ctx.request.body as Array<{ subscription_id: number, state: SubscriptionState }>
    for (const subscription of subscriptions) {
        await toggleSubscription(
            ctx.state.user!.id,
            subscription.subscription_id,
            subscription.state,
        )
    }
    ctx.body = await getUser(ctx.state.user!.id)
})

router.get('/:userId/journeys', async ctx => {
    const params = extractQueryParams(ctx.query, searchParamsSchema)
    ctx.body = await pagedEntrancesByUser(ctx.state.user!.id, params)
})

export default router
