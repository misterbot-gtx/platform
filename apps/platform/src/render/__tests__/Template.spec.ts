import { Variables } from '..'
import Project from '../../projects/Project'
import { User } from '../../users/User'
import { EmailTemplate, PushTemplate } from '../Template'

describe('Template', () => {
    describe('compile email', () => {
        test('link wrap enabled', async () => {
            const url = 'https://google.com'
            const template = EmailTemplate.fromJson({
                type: 'email',
                data: {
                    from: 'from',
                    subject: 'subject',
                    html: `<html><body><a href="${url}">link</a></body></html>`,
                },
            })
            const variables: Variables = {
                project: Project.fromJson({
                    link_wrap_email: true,
                }),
                user: User.fromJson({ id: 1 }),
                context: {
                    template_id: 1,
                    campaign_id: 1,
                    reference_id: '1',
                    subscription_id: 1,
                    reference_type: 'type',
                },
            }
            const compiled = template.compile(variables)
            expect(compiled).toMatchSnapshot()
        })

        test('link wrap disabled', async () => {
            const url = 'https://google.com'
            const template = EmailTemplate.fromJson({
                type: 'email',
                data: {
                    from: 'from',
                    subject: 'subject',
                    html: `<html><body><a href="${url}">link</a></body></html>`,
                },
            })
            const variables: Variables = {
                project: Project.fromJson({
                    link_wrap_email: false,
                }),
                user: User.fromJson({ id: 1 }),
                context: {
                    template_id: 1,
                    campaign_id: 1,
                    reference_id: '1',
                    subscription_id: 1,
                    reference_type: 'type',
                },
            }
            const compiled = template.compile(variables)
            expect(compiled).toMatchSnapshot()
        })
    })

    describe('compile push', () => {
        test('link wrap disabled', async () => {
            const url = 'https://google.com'
            const template = PushTemplate.fromJson({
                type: 'push',
                data: {
                    title: 'title',
                    topic: 'topic',
                    body: 'body',
                    url,
                    custom: {
                        key: 'value',
                    },
                },
            })
            const variables: Variables = {
                project: Project.fromJson({
                    link_wrap_push: false,
                }),
                user: User.fromJson({ id: 1 }),
                context: {
                    template_id: 1,
                    campaign_id: 1,
                    reference_id: '1',
                    subscription_id: 1,
                    reference_type: 'type',
                },
            }
            const compiled = template.compile(variables)
            expect(compiled).toMatchSnapshot()
            expect(compiled.custom.url).toBe(url)
        })

        test('link wrap enabled', async () => {
            const url = 'https://google.com'
            const template = PushTemplate.fromJson({
                type: 'push',
                data: {
                    title: 'title',
                    topic: 'topic',
                    body: 'body',
                    url,
                    custom: {
                        key: 'value',
                    },
                },
            })
            const variables: Variables = {
                project: Project.fromJson({
                    link_wrap_push: true,
                }),
                user: User.fromJson({ id: 1 }),
                context: {
                    template_id: 1,
                    campaign_id: 1,
                    reference_id: '1',
                    subscription_id: 1,
                    reference_type: 'type',
                },
            }
            const compiled = template.compile(variables)
            expect(compiled).toMatchSnapshot()
            expect(compiled.custom.url).not.toBe(url)
        })
    })
})
