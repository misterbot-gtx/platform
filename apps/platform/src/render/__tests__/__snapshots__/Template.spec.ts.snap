// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Template compile email link wrap disabled 1`] = `
Object {
  "from": "from",
  "html": "<html><body><a href=\\"https://google.com\\">link</a><img border=\\"0\\" width=\\"1\\" height=\\"1\\" alt=\\"\\" src=\\"https://parcelvoy.com/o?u=M8LRMWZ645&c=M8LRMWZ645&s=1\\" /></body></html>",
  "subject": "subject",
  "text": "link [https://google.com]",
}
`;

exports[`Template compile email link wrap enabled 1`] = `
Object {
  "from": "from",
  "html": "<html><body><a href=\\"https://parcelvoy.com/c?u=M8LRMWZ645&c=M8LRMWZ645&s=1&r=https%253A%252F%252Fgoogle.com\\">link</a><img border=\\"0\\" width=\\"1\\" height=\\"1\\" alt=\\"\\" src=\\"https://parcelvoy.com/o?u=M8LRMWZ645&c=M8LRMWZ645&s=1\\" /></body></html>",
  "subject": "subject",
  "text": "link [https://google.com]",
}
`;

exports[`Template compile push link wrap disabled 1`] = `
Object {
  "body": "body",
  "custom": Object {
    "key": "value",
    "url": "https://google.com",
  },
  "title": "title",
  "topic": "topic",
}
`;

exports[`Template compile push link wrap enabled 1`] = `
Object {
  "body": "body",
  "custom": Object {
    "key": "value",
    "url": "https://parcelvoy.com/c?u=M8LRMWZ645&c=M8LRMWZ645&s=1&r=https%253A%252F%252Fgoogle.com",
  },
  "title": "title",
  "topic": "topic",
}
`;
